tidy: move rustdoc js stuff into a tidy extra check

lolbinarycat opened the pull request:
Most of these were factored out of CI scripts, but `eslint` in particular was previously implemented with its own special cased logic.

A new option has been added to bootstrap, `build.tidy-extra-checks`, which serves as a default value for the `--extra-checks` flag.  This is mostly for the benefit of rustdoc js maintainers, but should also help bootstrap py maintainers.

Additionally, `--extra-checks=cpp` has been documented.

I'm not super happy with how long the extra check names are in comparison to the others (in particular `typecheck`), but I couldn't think of anything better (I didn't want to name it `tsc` on the off chance we want to switch to a different typechecking engine in the future).

It would be nice to convert the extra checks arg into a proper enum, both for warning on unknown values and to provide better shell completion.

r? @GuillaumeGomez 
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Fixes: https://github.com/rust-lang/rust/issues/144093

lolbinarycat committed:
tidy: running of eslint, tsc, and es-check are now an extra check

lolbinarycat committed:
update CI to use new tidy extra checks for rustdoc js

Kobzol commented on src/bootstrap/src/core/config/flags.rs:
```diff
@@ -382,7 +382,7 @@ pub enum Subcommand {
         #[arg(long)]
         /// comma-separated list of other files types to check (accepts py, py:lint,
-        /// py:fmt, shell)
+        /// py:fmt, shell, cpp, cpp:fmt, js, js:lint, js:typecheck, js:es-check)
```
Do we need to separate type checking and eslint? It's essentially the same category of "checking code" to me, at least in the world of JavaScript. Unless it runs for too long, I'd just collapse both into "lint" or "check". When you develop on JS, you'd set your `bootstrap.toml` file to contain both anyway, I imagine.

Kobzol commented on src/ci/docker/host-x86_64/mingw-check-1/Dockerfile:
```diff
@@ -55,6 +55,4 @@ ENV SCRIPT \
-           # Runs checks to ensure that there are no issues in our JS code.
-           es-check es2019 ../src/librustdoc/html/static/js/*.js && \
-           tsc --project ../src/librustdoc/html/static/js/tsconfig.json
+	   src/tools/tidy --extra-checks=js
```
Since this is now part of `tidy`, I think that we can move it out of this CI runner and just put it into `mingw-check-tidy`, to have centralized execution of all tidy checks on one place.

Kobzol commented on src/bootstrap/src/core/config/toml/build.rs:
```diff
@@ -69,6 +69,7 @@ define_config! {
         jobs: Option<u32> = "jobs",
         compiletest_diff_tool: Option<String> = "compiletest-diff-tool",
         compiletest_use_stage0_libtest: Option<bool> = "compiletest-use-stage0-libtest",
+        tidy_extra_checks: Option<String> = "tidy-extra-checks",
```
There should be a change tracker entry about this new option.

Kobzol reviewed (commented):
Thanks! Moving this from CI scripts to tidy is definitely an improvement. Left a few comments.

GuillaumeGomez commented on src/tools/tidy/src/ext_tool_checks.rs:
```diff
@@ -224,6 +248,18 @@ fn check_impl(
     }
 
+    if js_lint {
+        rustdoc_js::lint(librustdoc_path, tools_path, src_path)?;
```
Didn't we say that the checks would be run if the JS files were modified too? Having them never run unless an extra arguments is passed is very inconvenient.

the8472 commented on src/tools/tidy/src/main.rs:
```diff
@@ -171,7 +170,17 @@ fn main() {
 
-        check!(ext_tool_checks, &root_path, &output_directory, bless, extra_checks, pos_args);
+        check!(
+            ext_tool_checks,
```
Why is this being added to a single check module? AFAICS `ext_tools` is running external programs serially, so the wall-time of all those tools adds up. The `check!` macro exists to run things in threads so they can run in parallel to cut down on walltime.

lolbinarycat replied:
It's in `ext_tools` because that's where `--extra-checks` is parsed.  I definitely agree that the parsing logic and stuff could use a big upgrade, and that tools should be run concurrently, but I think that can be done within ext_tools.  It doesn't even technically need threads, just to spawn processes then wait for them all simultaneously.  I could get working on a system for that if you want.  It would probably involve making a struct or trait to represent an extra-check instead of having it be ad-hoc.  This would also be a good opportunity to refactor the parsing.

the8472 replied:
Well, the macro also limits concurrency to obey `-j`, so having thread per check is the easiest way to get that.

lolbinarycat replied:
So, should I inline every `--extra-check` as its own call to `check!`, probably doubling the size of `main()` in the process, or do you have a better idea?

the8472 replied:
Well, before doing more work I think it'd make sense to check how long it takes to run all the commands and compare that to the overall runtime of tidy. If it's the longest chain then parallelizing makes sense.

the8472 replied:
#143032 happens to have some numbers. We should probably move some of the longer-running tests to the front of main so that other ones get a chance to execute in parallel with them.

bors commented:
:umbrella: The latest upstream changes (presumably #142581) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

yotamofek commented on src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:
```diff
@@ -0,0 +1,142 @@
+    // Having the correct `eslint` version installed via `npm` isn't strictly necessary, since we're invoking it via `npx`,
+    // but this check allows the vast majority that is not working on the rustdoc frontend to avoid the penalty of running
+    // `eslint` in tidy. See also: https://github.com/rust-lang/rust/pull/142851
+    match get_eslint_version() {
```
I *do* think we can get rid of this check. I wanted to remove it in #142851 (in the first iteration, before I moved everything to be gated under the `--extra-checks` flag), but it would've made skipping `eslint` harder.
If we're going in the direction of only running `eslint` when needed, not just for everyone, then there's absolutely no reason to ask the user to explicitly install it, and not just use `npx` as it was intended to be used.
(just take care to run `npx eslint@<version we pinned>`)

yotamofek replied:
Yeah, but then most likely `npm` isn't either (unless someone is running a [really cursed environment](https://xkcd.com/1987/) [with `s/python/node/`] 😁)

RalfJung commented on src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:
```diff
@@ -0,0 +1,142 @@
+
+pub(super) fn typecheck(librustdoc_path: &Path) -> Result<(), super::Error> {
+    // use npx to ensure correct version
+    let mut child = Command::new("npx")
+        .arg("tsc")
+        .arg("-p")
+        .arg(librustdoc_path.join("html/static/js/tsconfig.json"))
+        .spawn()?;
```
So, uh, do all of these `npx` invocations basically mean "download random JS form the web and run it locally"? Without any kind of lockfile ensuring integrity of the download? That seems... bad?

yotamofek replied:
Well, `npx` will first look for an `eslint` installed with `npm install [-g]`, but otherwise yeah. See #142902 (which I self-assigned, but am waiting for this PR to be merged before I fix it).

yotamofek replied:
I mean, `npm install eslint@some.specific.version && npx eslint` is the same as `npx eslint@some.specific.version`,
in the sense that without a `package-lock.json` file (and an `--frozen` flag, I think?) both will run a certain version of `eslint`, but might run any semver-compatible version of the direct and transient deps.

RalfJung replied:
Yeah, I don't think we should do any of that without *explicit* consent from the user. Running arbitrary code without integrity checks is a complete no-go.

yotamofek replied:
Given the current state of things, not running the version check with `npm` just makes `tidy` a bit faster for whoever is already opting in to the `eslint` check, with this PR.

RalfJung replied:
> The state right now, before this PR, is that running tidy will ask you to run npm install eslint@some-version, and then will run that version of eslint the next time.

That is a way of asking for explicit consent, and very different from running `npm install` on the user's behalf.

> Yeah, but since npx won't be invoked unless the correct (i.e. rustc-repo-specified) version of eslint is installed, it has the same effect.

That is because of the check which you [suggested](https://github.com/rust-lang/rust/pull/142924#discussion_r2169821683) could be removed, right? Also, this comment here is attached to something doing `npx tsc`, not eslint, so -- I don't follow.

> whether we check if that version is already installed and then run it with npx, or just run it with npx eslint@that-version doesn't make that much of a difference IMHO.

It makes a huge difference whether you ask the user to do something vs whether you just do it without asking them. Asking for consent is different from just assuming consent!

yotamofek replied:
If we had that, I think that a comment about what it means to run `--extra-checks=eslint` would suffice, in terms of getting the user's permission to run `npx install --frozen eslint` (not sure if `--frozen` is the correct flag but consider it pseudo-shell)

lolbinarycat committed:
tidy: use a lockfile for js tools instead of npx

this makes us less vulnerable to MITM and supply chain attacks.

it also means that the CI scripts are no longer responsible for
tracking the versions of these tools.

it should also avoid the situation where local tsc and CI
disagree on the presense of errors due to them being different versions.

rust-log-analyzer commented:
The job **`mingw-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15934423731/job/44950992817) [(plain)](https://github.com/rust-lang/rust/commit/1e602f3ae86df192bb560fc65428797e6bdea6e5/checks/44950992817/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 3.161 Building wheels for collected packages: reuse
#15 3.162   Building wheel for reuse (pyproject.toml): started
#15 3.374   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.375   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 3.375   Stored in directory: /tmp/pip-ephem-wheel-cache-2mvf33nh/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 3.377 Successfully built reuse
#15 3.378 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.786 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.787 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 DONE 3.9s
---
Checking tidy rustdoc_json...
No error code explanation was removed!
`rustdoc-json-types` was not modified.
tidy: Skipping binary file check, read-only filesystem
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /checkout/obj/package.json
npm ERR! errno -2
npm ERR! enoent Could not read package.json: Error: ENOENT: no such file or directory, open '/checkout/obj/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-06-27T20_06_45_554Z-debug-0.log
tidy error: checks with external tool 'npm install failed' failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:00:54
  local time: Fri Jun 27 20:06:45 UTC 2025
  network time: Fri, 27 Jun 2025 20:06:45 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15935404778/job/44954095687) [(plain)](https://github.com/rust-lang/rust/commit/ec37d894484ffc01c9c7296e79366f622b5fbb04/checks/44954095687/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.890 Building wheels for collected packages: reuse
#15 2.891   Building wheel for reuse (pyproject.toml): started
#15 3.106   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.107   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 3.107   Stored in directory: /tmp/pip-ephem-wheel-cache-9d6ei4fq/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 3.110 Successfully built reuse
#15 3.110 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.518 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.518 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 DONE 3.6s
---
Checking tidy rustdoc_json...
`rustdoc-json-types` was not modified.
No error code explanation was removed!
tidy: Skipping binary file check, read-only filesystem
npm ERR! code ENOENT
npm ERR! syscall mkdir
npm ERR! path /checkout/node_modules
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/checkout/node_modules'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-06-27T21_06_24_376Z-debug-0.log
some tidy checks failed
tidy error: checks with external tool 'npm install failed' failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:00:57
  local time: Fri Jun 27 21:06:24 UTC 2025
  network time: Fri, 27 Jun 2025 21:06:24 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Kobzol commented on bootstrap.example.toml:
```diff
@@ -467,6 +467,15 @@
 # Whether to use the precompiled stage0 libtest with compiletest.
 #build.compiletest-use-stage0-libtest = true
 
+# Default value for the `--extra-checks` flag of tidy.
```
This PR is becoming quite large, and this looks quite self-contained (and useful on its own). Could you please land it separately?

bors commented:
:umbrella: The latest upstream changes (presumably #143254) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16202181063/job/45743765075) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45743765075) [(plain)](https://github.com/rust-lang/rust/commit/9737648cd9695793a609d4d6d5a54f2231d5db4c/checks/45743765075/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.752 Building wheels for collected packages: reuse
#15 2.753   Building wheel for reuse (pyproject.toml): started
#15 2.967   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 2.968   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 2.968   Stored in directory: /tmp/pip-ephem-wheel-cache-dncikjpj/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 2.970 Successfully built reuse
#15 2.971 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.370 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.370 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 3.908 Collecting virtualenv
---
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
npm ERR! code ENOENT
npm ERR! syscall mkdir
npm ERR! path /checkout/node_modules
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/checkout/node_modules'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-07-10T17_51_52_287Z-debug-0.log
tidy error: checks with external tool 'npm install failed' failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:24
  local time: Thu Jul 10 17:51:52 UTC 2025
  network time: Thu, 10 Jul 2025 17:51:52 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16202697061/job/45745598822) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45745598822) [(plain)](https://github.com/rust-lang/rust/commit/f50898d7f874cf72608e3d47206f667f1cb72c47/checks/45745598822/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.793 Building wheels for collected packages: reuse
#15 2.794   Building wheel for reuse (pyproject.toml): started
#15 3.006   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.006   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 3.007   Stored in directory: /tmp/pip-ephem-wheel-cache-7jt0w4u5/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 3.009 Successfully built reuse
#15 3.009 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.402 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.403 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 3.929 Collecting virtualenv
---
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
npm ERR! code ENOENT
npm ERR! syscall mkdir
npm ERR! path /checkout/node_modules
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/checkout/node_modules'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-07-10T18_19_33_453Z-debug-0.log
tidy error: checks with external tool 'npm install' failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:20
  local time: Thu Jul 10 18:19:33 UTC 2025
  network time: Thu, 10 Jul 2025 18:19:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lolbinarycat commented on package.json:
Is there a good way to make npm do that? it's also worth noting that `rustdoc-gui-test` seems to create it in the dir root, which is the entire reason I put it (and the package files) there.

bjorn3 replied:
You could copy the `package.json` and `package-lock.json` files to the build directory and then invoke `npm ci` from the build directory.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16206491694/job/45758015577) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45758015577) [(plain)](https://github.com/rust-lang/rust/commit/f303b866d4b15391e81b37321a98f1de9568c262/checks/45758015577/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.849 Building wheels for collected packages: reuse
#15 2.850   Building wheel for reuse (pyproject.toml): started
#15 3.067   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.068   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 3.069   Stored in directory: /tmp/pip-ephem-wheel-cache-xzxyx9ht/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 3.071 Successfully built reuse
#15 3.071 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.472 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.472 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 4.030 Collecting virtualenv
---
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
some tidy checks failed
tidy error: IO error: No such file or directory (os error 2)
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:20
  local time: Thu Jul 10 21:41:49 UTC 2025
  network time: Thu, 10 Jul 2025 21:41:49 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16207031334/job/45759655643) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45759655643) [(plain)](https://github.com/rust-lang/rust/commit/fe7e0c5f33f2c51fdaca741b9829ef9b0ffb566b/checks/45759655643/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.833 Building wheels for collected packages: reuse
#15 2.834   Building wheel for reuse (pyproject.toml): started
#15 3.046   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.047   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 3.047   Stored in directory: /tmp/pip-ephem-wheel-cache-nljjzyqs/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 3.050 Successfully built reuse
#15 3.050 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.439 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.439 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 4.000 Collecting virtualenv
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
error[E0308]: mismatched types
  --> src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:49:31
   |
49 |     let mut child = spawn_cmd(cmd)?;
   |                     --------- ^^^ expected `&mut Command`, found `Command`
   |                     |
   |                     arguments to this function are incorrect
   |
note: function defined here
  --> src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:18:4
   |
18 | fn spawn_cmd(cmd: &mut Command) -> Result<Child, io::Error> {
   |    ^^^^^^^^^ -----------------
help: consider mutably borrowing here
   |
49 |     let mut child = spawn_cmd(&mut cmd)?;
   |                               ++++

error[E0308]: mismatched types
   --> src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:139:21
    |
139 |     match spawn_cmd(cmd)?.wait() {
    |           --------- ^^^ expected `&mut Command`, found `Command`
    |           |
    |           arguments to this function are incorrect
    |
note: function defined here
   --> src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:18:4
    |
18  | fn spawn_cmd(cmd: &mut Command) -> Result<Child, io::Error> {
    |    ^^^^^^^^^ -----------------
help: consider mutably borrowing here
    |
139 |     match spawn_cmd(&mut cmd)?.wait() {
    |                     ++++

error: unused variable: `eslint_version`
  --> src/tools/tidy/src/ext_tool_checks/rustdoc_js.rs:96:9
   |
96 |     let eslint_version = match std::fs::read_to_string(&eslint_version_path) {
   |         ^^^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_eslint_version`
   |
   = note: `-D unused-variables` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unused_variables)]`

For more information about this error, try `rustc --explain E0308`.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16207225337/job/45760260287) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45760260287) [(plain)](https://github.com/rust-lang/rust/commit/d6e1a36406a480b8ed37e52c861f8c96c0804a09/checks/45760260287/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 3.058 Building wheels for collected packages: reuse
#15 3.059   Building wheel for reuse (pyproject.toml): started
#15 3.272   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.273   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 3.273   Stored in directory: /tmp/pip-ephem-wheel-cache-dmn7s6id/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 3.276 Successfully built reuse
#15 3.276 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.671 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.671 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 4.247 Collecting virtualenv
```

</details>

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16208491854/job/45763953264) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45763953264) [(plain)](https://github.com/rust-lang/rust/commit/98204a1f5276f413a2473999e3e0b2f790380cfe/checks/45763953264/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.746 Building wheels for collected packages: reuse
#15 2.747   Building wheel for reuse (pyproject.toml): started
#15 2.959   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 2.960   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 2.961   Stored in directory: /tmp/pip-ephem-wheel-cache-e846zqnz/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 2.963 Successfully built reuse
#15 2.963 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.372 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.372 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 DONE 3.5s
---


# RECOMMENDATIONS

* Fix missing copyright/licensing information: For one or more files, the tool
  cannot find copyright and/or licensing information. You typically do this by
  adding 'SPDX-FileCopyrightText' and 'SPDX-License-Identifier' tags to each
  file. The tutorial explains additional ways to do this:
  <https://reuse.software/tutorial/>

  local time: Fri Jul 11 00:29:37 UTC 2025
  network time: Fri, 11 Jul 2025 00:29:37 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lolbinarycat committed:
REUSE.toml: add new package.json and package-lock.json

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16209034847/job/45765595847) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45765595847) [(plain)](https://github.com/rust-lang/rust/commit/52ae02c887e1029ff0b07401cd03cf78deab7fab/checks/45765595847/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16210114076/job/45768526617) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45768526617) [(plain)](https://github.com/rust-lang/rust/commit/acd3f5038fc96d60066cc88bdb8fbcd57b25b716/checks/45768526617/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 2.755 Building wheels for collected packages: reuse
#15 2.756   Building wheel for reuse (pyproject.toml): started
#15 2.971   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 2.971   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#15 2.972   Stored in directory: /tmp/pip-ephem-wheel-cache-x862_mqw/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#15 2.974 Successfully built reuse
#15 2.975 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#15 3.378 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#15 3.378 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#15 DONE 3.5s
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16210459896/job/45769418083) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45769418083) [(plain)](https://github.com/rust-lang/rust/commit/1db5d87c974937dfd6a9ea991eb047e1bd64e422/checks/45769418083/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING] core::build_steps::test::Rustfmt { stage: 2, host: x86_64-unknown-linux-gnu } -- 54.813
Build completed successfully in 0:02:47
+ head -n 1 /tmp/browser-ui-test.version
+ npm install browser-ui-test@0.21.1 --unsafe-perm=true
npm ERR! code ENOENT
npm ERR! syscall mkdir
npm ERR! path /checkout/node_modules
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/checkout/node_modules'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-07-11T03_39_06_508Z-debug-0.log
  local time: Fri Jul 11 03:39:07 UTC 2025
  network time: Fri, 11 Jul 2025 03:39:07 GMT
##[error]Process completed with exit code 254.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16225674633/job/45817053180) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45817053180) [(plain)](https://github.com/rust-lang/rust/commit/44def3a0d14afde29f95a65c334df9884c6e6f23/checks/45817053180/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING] core::build_steps::test::Rustfmt { stage: 2, host: x86_64-unknown-linux-gnu } -- 56.851
Build completed successfully in 0:02:52
+ head -n 1 /tmp/browser-ui-test.version
+ npm install browser-ui-test@0.21.1 --unsafe-perm=true
npm ERR! code ENOENT
npm ERR! syscall mkdir
npm ERR! path /checkout/node_modules
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/checkout/node_modules'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-07-11T18_11_12_636Z-debug-0.log
  local time: Fri Jul 11 18:11:13 UTC 2025
  network time: Fri, 11 Jul 2025 18:11:13 GMT
##[error]Process completed with exit code 254.
Post job cleanup.
```

</details>

GuillaumeGomez commented on src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile:
```diff
@@ -100,6 +100,6 @@ COPY scripts/shared.sh /scripts/
 # the local version of the package is different than the one used by the CI.
 ENV SCRIPT /tmp/checktools.sh ../x.py && \
-  npm install browser-ui-test@$(head -n 1 /tmp/browser-ui-test.version) --unsafe-perm=true && \
+  npm install --no-package-lock browser-ui-test@$(head -n 1 /tmp/browser-ui-test.version) --unsafe-perm=true && \
```
Now that it's in the `package.json` file, do we still need the `head -n 1 /tmp/browser-ui-test.version` part?

lolbinarycat replied:
I'm currently opening a separate PR to refactor how `browser-ui-test` gets installed, since this method seems quite fragile.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16228580147/job/45826261595) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45826261595) [(plain)](https://github.com/rust-lang/rust/commit/cd24e997079ffd7e7411993777052d2578dfa2b8/checks/45826261595/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 52.346 seconds
[TIMING] core::build_steps::test::Rustfmt { stage: 2, host: x86_64-unknown-linux-gnu } -- 52.347
Build completed successfully in 0:02:41
+ head -n 1 /tmp/browser-ui-test.version
+ npm install --no-package-lock browser-ui-test@0.21.1 --unsafe-perm=true
npm ERR! code ENOENT
npm ERR! syscall mkdir
npm ERR! path /checkout/node_modules
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/checkout/node_modules'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-07-11T20_53_40_292Z-debug-0.log
  local time: Fri Jul 11 20:53:55 UTC 2025
  network time: Fri, 11 Jul 2025 20:53:55 GMT
##[error]Process completed with exit code 254.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #143919) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lolbinarycat commented:
This logic should also be updated to make it respect `build.npm`.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16301231779/job/46036199238) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46036199238) [(plain)](https://github.com/rust-lang/rust/commit/b339aecf6a109d4b4db2e32df7ade86863dcb636/checks/46036199238/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
downloading https://static.rust-lang.org/dist/2025-06-24/rustfmt-nightly-aarch64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2025-06-24/rustfmt-nightly-aarch64-unknown-linux-gnu.tar.xz to /checkout/obj/build/aarch64-unknown-linux-gnu/rustfmt
downloading https://static.rust-lang.org/dist/2025-06-24/rustc-nightly-aarch64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2025-06-24/rustc-nightly-aarch64-unknown-linux-gnu.tar.xz to /checkout/obj/build/aarch64-unknown-linux-gnu/rustfmt
x.py completions were changed; run `x.py run generate-completions` to update them
Build completed unsuccessfully in 0:00:06
  local time: Tue Jul 15 18:34:56 UTC 2025
  network time: Tue, 15 Jul 2025 18:34:56 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16392060015/job/46319342099) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46319342099) [(plain)](https://github.com/rust-lang/rust/commit/cca02b0780ed12e0bc41457d244e8942ff4c9388/checks/46319342099/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
extracting /checkout/obj/build/cache/2025-06-24/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
downloading https://static.rust-lang.org/dist/2025-06-24/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2025-06-24/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.215
x.py completions were changed; run `x.py run generate-completions` to update them
Build completed unsuccessfully in 0:00:14
  local time: Sat Jul 19 19:41:04 UTC 2025
  network time: Sat, 19 Jul 2025 19:41:04 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lolbinarycat committed:
update completions

lolbinarycat committed:
integrate build_helper::npm into js checks and package.json usage

Kobzol reviewed (commented):
Looks great! Please resolve the FIXME with passing `build.npm` from bootstrap, and then you can r=me and Guillaume.

Kobzol commented:
@bors delegate+

bors commented:
:v: @lolbinarycat, you can now approve this pull request!

If @Kobzol told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@Kobzol`
<!-- homu: {"type":"Delegated","delegator":"Kobzol","delegate":"lolbinarycat"} -->

GuillaumeGomez commented:
Just a note: since @Kobzol already approved the PR minus the change, you can r+ the PR as soon as the change is done with `rollup` so it can be merged super quickly.
